### PR TITLE
A much simpler way to fix finetune

### PR DESCRIPTION
**Pull request by zhongwen**
A much simpler way to fix Issue #147.

It might be unnecessary to have large modifications on pooling layers, which will make the old models not suitable for fine-tuning, and add new pooling layers with the save_params option. The pooling layers with `LoadModel` and `SaveModel` might be a bit confused.

Please consider this and may revert previous fixes if it is possible.

### CODE REVIEW

**Comment by winstywang** [comment]
Does it work if I skip some layers?

**Comment by zhongwen** [comment]
just copy layers by layer names

**Comment by zhongwen** [comment]
oh I know what you mean. I think it is wield if skipping some layers. anyway. you may init the old net but not call connect layers, but not the way to modify pooling layers i think. sorry i have to go to bed now.

**Comment by zhulinchao** [comment]
Hi, in my understanding, what zhongwen means is that the bug is caused by `InitConnection` between layers in `old_net`, as most params in `old_net` are in default value. I think the following path can fix the bug as well:
```diff
---
 src/nnet/neural_net-inl.hpp |   14 +++++++++++++-
 src/nnet/nnet_impl-inl.hpp  |    6 +++---
 2 files changed, 16 insertions(+), 4 deletions(-)

diff --git a/src/nnet/neural_net-inl.hpp b/src/nnet/neural_net-inl.hpp
index 7042c8a..1a26086 100644
--- a/src/nnet/neural_net-inl.hpp
+++ b/src/nnet/neural_net-inl.hpp

@@ -84,6 +84,18 @@ struct NeuralNet {
       }
     }
   }
+  inline void LoadModel_S1(utils::IStream &fi) {
+    this->FreeSpace();
+    this->InitNet();
+    this->ConfigConntions();
+    for (size_t i = 0; i < connections.size(); ++i) {
+      if (connections[i].type != layer::kSharedLayer) {
+        connections[i].SetStream(stream);
+        connections[i].layer->LoadModel(fi);
+      }
+    }
+  }
+
   /*! \brief load model from stream */
   inline void LoadModel(utils::IStream &fi) {
     this->FreeSpace();
diff --git a/src/nnet/nnet_impl-inl.hpp b/src/nnet/nnet_impl-inl.hpp
index 6aa8c48..fdaed6d 100644
--- a/src/nnet/nnet_impl-inl.hpp
+++ b/src/nnet/nnet_impl-inl.hpp
@@ -123,11 +123,11 @@ class CXXNetThreadTrainer : public INetTrainer {
     old_cfg.LoadNet(fi);
     fi.Read(&epoch_counter, sizeof(epoch_counter));
     epoch_counter = 0;
-    NeuralNet<cpu> old_net(old_cfg, 0, 0, NULL);
+    NeuralNet<cpu> old_net(old_cfg, 1, 0, NULL);
     std::string old_model;
     fi.Read(&old_model);
     utils::MemoryBufferStream os(&old_model);
-    old_net.LoadModel(os);
+    old_net.LoadModel_S1(os);

     // Compare original net and current net
     for (index_t i = 0; i < old_cfg.layers.size(); ++i){
--
1.7.4.1
```

**Comment by winstywang** [comment]
It is quite common actually, considering the case that contains a split layer or the DSN case. They both need the case that skipping some layers.

**Comment by winstywang** [comment]
@zhulinchao Can you make sure the fix work?

**Comment by zhongwen** [comment]
Split layer is OK, consider the structure in batch normalization for example, the one to initialize from scratch is only the final layer.

The only exception is auxiliary loss style, which has multiple "classifiers" branches. In this case, the layers removed in fine-tuning stage are not only the last few consecutive layers.

Well, I haven't read a paper using fine-tuning with auxiliary loss functions. Though many Google guys use GoogLeNet, but since Batch Normalization appears, the auxiliary loss becomes unnecessary. 

Yes, there's a recent paper using DSN on ImageNet which achieves good results. But I do think the most common and popular nets used in fine-tuning are AlexNet, ZFNet, and VGGNet, all are simple one-tower models.

Nonetheless, I do agree that the way to parse the old model only but not continue to build the old_net (patch @zhulinchao gave) is more robust. I just think the way which makes the trained models not working for fine-tuning might be unreasonable, just think about it, if you have been training a model for weeks and turn to fine-tuning finally, you found it buggy in previous version, and you'd like to merge from the latest version, in the end, you find that it tells you that the model you trained for weeks cannot be used for fine-tuning. The way to modify pooling layers is unnecessary.

Please test his patch on your DSN fine-tuning conf, we do not have one.

**Comment by winstywang** [comment]
I see, I will test it today after I finish the caffe converter.

**Comment by zhongwen** [comment]
I have written one caffeconverter (see below), though my converter only consider conv layers and fc layers, a bit specific for VGG-16. I have tested it on CNN-M as well, though the mean image is not considered in the following scripts.

https://gist.github.com/zhongwen/05829a531e5236ddb8ec

**Comment by winstywang** [comment]
Thanks, we have considered the mean file in our converter. Almost done :)

**Comment by antinucleon** [comment]
I used to make one but there it doesn't work correctly. Also In found way to convert mean.  Maybe we can quickly fix it today or tomorrow. https://github.com/antinucleon/cxxnet/blob/caffe_converter/tools/caffe_model_2_cxxnet.py

**Comment by winstywang** [comment]
fixed
